### DISCUSSION TITLE
Relate using external_identifier or other non-primary keys

### DISCUSSION THREAD

**Discussion by jeremybradbury** [directus/directus#12489] [Feature Requests]
The problems is Directus' assumption that all unique keys are an `id` or primary key.

- I add a foreign key to (or from) external_identifier
- the relationship appears in the graph... win!
- or is it?...

Directus has ignored the foreign keys and assumes all relations go to an object's primary key. It tries to match my other id with directus_users.id instead. 

This breaks a lot of common patterns, like simple 1:1s on a second, optional unique index.

The solution is to allow a relationship to any unique index (not just the primary key), there are fields in the relations table to handle this already, but they are ignored when used.

implementation ideas:

1.  allow me to add a relationship to external_identifier with a unique index (one user account, one profile)
1.  allow me to add a relationship from external_identifier to another collection (one profile, many user accounts)
or
1.  allow me to add a relationship "field", virtual (we only need a collection name & the 2 keys) or actual (we could put a PK here after matching the ext id), that is tied to the external_identifier 

It's a very very useful field if you can use it for relationships. The same is true in many other tables and patterns that have multiple unique indexes. many are secondary indexes, to provide nullable (pks cannot be nullable) add-ons to any object.

For example email is likely a unique index for user accounts. And even if it isn't, we should be able to relate that to another table, one that may perhaps also not use email as the primary key, just 2 matching unique values to join 2 tables on. 

We really only need one unique field list anyway and nulls don't break joins or where clauses, only primary keys.

See here for some screenshots of the admin UI lying/mistaken/assuming about the relationship I added from the db and more detail: https://github.com/directus/directus/discussions/12091

**Comment by jeremybradbury** [comment]
I have one workaround for this: 

Making a table with a string primary key so it can match the string datatype of external_identifier.

To me having a VARCHAR for a primary key is unacceptable (esp a 255). I can get my PK down to VARCHAR(15) since the third party uses bigints, but still the index is massive, and likely will need to become 36 adding other providers.

This also requires directus admin users to add the ids themselves when creating a new record.

PKs should be UUIDs almost always and bigints for small lists.

anyway, this pattern currently allows me to attach an FK to external_identifier, so that my users_me graph connects to the user's data....

But it's incredibly bad DBA/Design just to play nice with Directus.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
